pyo3-macros-backend/src/pyclass.rs

```diff
@@ -383,242 +371,127 @@ fn impl_class(
-                        ::std::option::Option::Some(::pyo3::class::impl_::free_with_freelist::<#cls>)
-                    }
-                }
-            }
-        });
+    let py_class_impl = PyClassImplsBuilder::new(cls, attr, methods_type)
+        .doc(doc)
+        .impl_all();
 
     let descriptors = impl_descriptors(cls, field_options)?;
 
-    // insert space for weak ref
-    let weakref = if attr.has_weaklist {
-        quote! { ::pyo3::pyclass_slots::PyClassWeakRefSlot }
-    } else if attr.has_extends {
-        quote! { <Self::BaseType as ::pyo3::class::impl_::PyClassBaseType>::WeakRef }
-    } else {
-        quote! { ::pyo3::pyclass_slots::PyClassDummySlot }
-    };
-    let dict = if attr.has_dict {
-        quote! { ::pyo3::pyclass_slots::PyClassDictSlot }
-    } else if attr.has_extends {
-        quote! { <Self::BaseType as ::pyo3::class::impl_::PyClassBaseType>::Dict }
-    } else {
-        quote! { ::pyo3::pyclass_slots::PyClassDummySlot }
-    };
-    let module = if let Some(m) = &attr.module {
-        quote! { ::std::option::Option::Some(#m) }
-    } else {
-        quote! { ::std::option::Option::None }
-    };
+    Ok(quote! {
+        #pytypeinfo_impl
 
-    // Enforce at compile time that PyGCProtocol is implemented
-    let gc_impl = if attr.is_gc {
-        let closure_name = format!("__assertion_closure_{}", cls);
-        let closure_token = syn::Ident::new(&closure_name, Span::call_site());
-        quote! {
-            fn #closure_token() {
-                use ::pyo3::class;
+        #py_class_impl
 
-                fn _assert_implements_protocol<'p, T: ::pyo3::class::PyGCProtocol<'p>>() {}
-                _assert_implements_protocol::<#cls>();
-            }
-        }
-    } else {
-        quote! {}
-    };
-
-    let (impl_inventory, for_each_py_method) = match methods_type {
-        PyClassMethodsType::Specialization => (None, quote! { visitor(collector.py_methods()); }),
-        PyClassMethodsType::Inventory => (
-            Some(impl_methods_inventory(cls)),
-            quote! {
-                for inventory in ::pyo3::inventory::iter::<<Self as ::pyo3::class::impl_::HasMethodsInventory>::Methods>() {
-                    visitor(::pyo3::class::impl_::PyMethodsInventory::methods(inventory));
-                }
-            },
-        ),
-    };
+        #descriptors
+    })
+}
 
-    let methods_protos = match methods_type {
-        PyClassMethodsType::Specialization => {
-            quote! { visitor(collector.methods_protocol_slots()); }
-        }
-        PyClassMethodsType::Inventory => {
-            quote! {
-                for inventory in ::pyo3::inventory::iter::<<Self as ::pyo3::class::impl_::HasMethodsInventory>::Methods>() {
-                    visitor(::pyo3::class::impl_::PyMethodsInventory::slots(inventory));
-                }
-            }
-        }
-    };
+struct VariantPyO3<'a> {
```

Comment:
Perhaps this name can be more descriptive:

```suggestion
struct PyClassEnumVariant<'a> {
```